### ISSUE TITLE
[BUG] Sudden high memory consumption by mx.eval

### ISSUE THREAD

**Issue by StefanS-prog** [ml-explore/mlx#2305]
**Describe the bug**
The gradient accumulation implementation with mx.eval causes a sudden high memory usage of > 10 GB. Before mx.eval it is around 600 MB which is reasonable.

**To Reproduce**
Just execute the complete program and observe the memory usage > 10 GB. Move the endless loop above mx.eval and observe again the memory usage of ca. 600 MB.
Include code snippet
```
# optimize
optimizers = model.configure_optimizers(weight_decay=0.1, learning_rate=6e-4)
grad_accum_steps = mx.array(grad_accum_steps)

def loss_fn(model, x, y):
    logits = model(x)
    loss = nn.losses.cross_entropy(logits.reshape((-1, logits.shape[-1])), y.reshape((-1)), reduction='mean')
    loss /= grad_accum_steps
    return loss

@mx.compile
def split_grads(grads):
    grads = tree_flatten(grads)
    weight_grads = [(k, p) for k, p in grads if p.ndim == 2]
    bias_grads = [(k, p) for k, p in grads if p.ndim == 1]
    weight_grads = tree_unflatten(weight_grads)
    bias_grads = tree_unflatten(bias_grads)
    return weight_grads, bias_grads

@partial(mx.compile, inputs=[model.state], outputs=[model.state])
def comp_loss_and_grads(x, y):
    loss_and_grad_fn = nn.value_and_grad(model, loss_fn)
    loss, grads = loss_and_grad_fn(model, x, y)
    return loss, grads

state = [model.state] + [optimizer.state for optimizer in optimizers]

@partial(mx.compile, inputs=state, outputs=state)
def update(grads, lr):
    grads, norm = optim.clip_grad_norm(grads, 1.0)
    for optimizer in optimizers:
        optimizer.learning_rate = lr
    weight_grads, bias_grads = split_grads(grads)
    optimizers[0].update(model, bias_grads)
    optimizers[1].update(model, weight_grads)
    return norm

for step in range(max_steps):
    t0 = time.time()
    for micro_step in range(grad_accum_steps.item()):
        x, y = train_loader.next_batch()
        #with torch.autocast(device_type=device, dtype=torch.bfloat16):
        if micro_step == 0:
            loss, grads = comp_loss_and_grads(x, y)
        else:
            d_loss, d_grads = comp_loss_and_grads(x, y)
            loss += d_loss
            grads = tree_map(lambda grads, d_grads: grads + d_grads, grads, d_grads)
        mx.eval(loss, grads)
        while True:
                pass
        print(micro_step)

    print(grads)
    import sys; sys.exit(0)
```

**Expected behavior**
The memory consumption should stay roughly the same at ca. 600 MB.

**Desktop (please complete the following information):**
 - OS Version: 15.5
 - Version: 24F74

**Additional context**
I was following along Andrej Karpathy's nanoGPT development and trying to port it to MLX.

**Comment by FL33TW00D** [comment]
@StefanS-prog MLX is a lazy framework, nothing happens until `mlx.eval` is performed. Learn more here: https://ml-explore.github.io/mlx/build/html/usage/lazy_evaluation.html

**Comment by StefanS-prog** [comment]
@FL33TW00D That I've read. But that does not explain the jump in used memory from ca. 600 MB to ca. 10 GB. After mlx.eval I expect it still to be at ca. 600 MB. What also happens after a few more iterations is that the physical RAM and swap space are exhausted, macOS locks up and a watchdog timeout causes a reboot.

**Comment by awni** [comment]
Can you provide the full code (or some code that we can run) that reproduces the issue you are seeing? Without that it's pretty hard to say if this is expected memory use or something fishy is going on.

**Comment by StefanS-prog** [comment]
@awni I have created the repository [MLX-gpt](https://github.com/StefanS-prog/MLX-gpt) with the complete file and the input file being used for training. You can also remove the infinite loop in the file. Then my Mac mini M2 Pro reboots around iteration 60.

**Comment by zcbenz** [comment]
To answer some questions in this issue:

The memory consumption jumped after `mx.eval` because that's when the memory were actually allocated, i.e. when you run something like `c = a + b` the memory for `c` would not be allocated until you call `mx.eval(c)`.

The memory consumption did not go down was because of the buffer cache — the memory of destroyed arrays were kept in a buffer pool and wouldn't be freed until you call `clear_cache()`. This is because memory allocation is slow and performance would be heavily penalized without cache. 

Since the repo is gone we are not able to look into what happened, and I tend to think it was just expected (though un-intuitive) behavior, so I'm closing this issue.
